File: src/main/kotlin/me/snwy/oasis/OasisFunction.kt
package me.snwy.oasis

class OasisFunction(@JvmField val declaration: Func, @JvmField val closure: Environment) : OasisCallable {
    override fun call(interpreter: Interpreter, arguments: List<Any?>): Any? {
        val environment = Environment(closure)
        if (arguments.size < arity()) throw RuntimeError(line, "Expected ${arity()} arguments, got ${arguments.size}")
        (0 until declaration.operands.size).map {
            environment.define(
                declaration.operands[it].lexeme.hashCode(),
                arguments[it]
            )
        }
        try {
            interpreter.executeBlock(declaration.body, environment)
        } catch (returnValue: Return) {
            return returnValue.value
        }
        return null
    }

    override fun arity(): Int {
        return declaration.operands.size
    }

    override fun toString(): String {
        return "me.snwy.OasisFunction(${declaration.operands.map { x -> x.lexeme }})"
    }
}

File: oasis_function.go
package main

type OasisCallable interface {
	Call(vm *VM, args []interface{}) interface{}
	Arity() int
}

type OasisFunction struct {
	Args    []string
	Body    []Instruction
	Closure *OasisEnvironment
}

func (f OasisFunction) Call(vm *VM, args []interface{}) interface{} {
	env := NewEnvironment(f.Closure)
	for i := 0; i < len(f.Args); i++ {
		env.Values[f.Args[i]] = args[i]
	}
	var fnVm = NewVM(f.Body, env, true)
	fnVm.Run()
	if fnVm.IteratorExhausted && vm != nil {
		vm.IteratorExhausted = true
	}
	return fnVm.ReturnValue
}

func (f OasisFunction) Arity() int {
	return len(f.Args)
}

type NativeFunction struct {
	Fn   func(vm *VM, args []interface{}) interface{}
	Args int
}

func (f *NativeFunction) Call(vm *VM, args []interface{}) interface{} {
	return f.Fn(vm, args)
}

func (f *NativeFunction) Arity() int {
	return f.Args
}
